README
Astro::Montenbruck

Library of astronomical calculations, based on "Astronomy On The Personal
Computer" by O.Montenbruck and T.Phleger, Fourth Edition, Springer-Verlag, 2000.

The main purpose of the library is to calculate positions of the Sun, the Moon,
and the planets with precision that is approximately the same as that found in
astronomical yearbooks. Other modules contain time-related routines, coordinates
conversions, calculation of the ecliptic obliquity and nutation, etc. Over time,
the range of utility functions will grow.

Partially it overlaps some code which already exists in CPAN and elsewhere.
For instance, there is a Perl wrapper for Swiss Ephemeris,
http://www.astrotexte.ch/sources/SwissEph.html. Swiss Ephemeris is fast and
precise C library. Unfortunately, it lacks portability and convenient license.
So, it is not easy for a layman to customize it for her custom application, be
it an online lunar calendar, or tool for amateur sky observations.

The present library is an attempt to find a middle-ground between precision on
the one hand and compact, well organized code on the other. I tried to follow the
best practices of modern Perl programming.

PRECISION

As the book authors state in Introduction to the 4-th edition,
"The errors in the fundamental routines for determining the coordinates of the
Sun, the Moon, and the planets amount to about 1″-3″.


INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build installdeps
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command from the parent directory of the library:

   perldoc Astro::Montenbruck

You can also generate local HTML documentation with

  perl script/createdocs.pl

Documentation files will be installed to docs directory.

USAGE

script/ directory contains examples of the library usage. They will be extended
over time.

* planpos.pl — positions of Sun, Moon and the planets
* riseset.pl — rises and sets of celestial objects 
* phases.pl — lunar phases
* rst_almanac.pl — rises/sets/transits events for a range of dates 
* solequ.pl — solstices and equinoxes
* mercury.pl — very simple demo of calculating Mercury position 
* moon_almanac — rises, sets of the Moon, its positions and lunar phase circumstances for a range of dates

For example, to display current planetary positions, type:

$ perl script/planpos.pl

For list of available options. type:

$ perl script/planpos.pl --help

Most of the scripts have --help option for a short list of options and --man for an extended description.  


HOW TO CONTRIBUTE

You may contribute to the project by many different ways, starting from refining and correcting its documentation,
especially if you are a native English speaker, and ending with improving the code base. Any kind of testing and
suggestions are welcome. 

You may follow the standard Github procedures or, in case you are not comfortable with them, just send your suggestions 
to the author by email.



LICENSE AND COPYRIGHT

Copyright (C) 2010-2022 Sergey Krushinsky

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (1.0). You may obtain a
copy of the full license at:

https://dev.perl.org/licenses/artistic.html

Aggregation of this Package with a commercial distribution is always
permitted provided that the use of this Package is embedded; that is,
when no overt attempt is made to make this Package's interfaces visible
to the end user of the commercial distribution. Such use shall not be
construed as a distribution of this Package.

The name of the Copyright Holder may not be used to endorse or promote
products derived from this software without specific prior written
permission.

THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.


Changes
Revision history for AstroScript

2019-06-18 - script ephemeris.pl replaced with planpos.pl.
             floating point error in tests fixed.
             Helpers module moved from 'lib' to 'script'
2019-06-20 - floating point error in test fixed

2019-08-10 - module for calculating rise, set and transit events with example
             script.
2020-12-20 - ability to use decimal lat/lon input

2020-12-21 - Lunar phases module and script

2020-12-26 - Solstices and Equinoxes 
             Aparent position of the Sun

2020-12-31 - Apparent positions of the objects, Planet interface changed
             Numerous bugfixes

2021-01-01 - Demo script for computing solstices / Equinoxes
             Utility modules renamed to avoid CPAN authorization issues

2021-01-05 - List::MoreUtils dependency added to build prerequisites

2021-01-09 - Routines for calculating rise/set/transit times rewritten. 
             Algorithms from J.Meeus book substituted with the alternatives,
             found in Montenbruck & Phleger's "Astronomy on the PC".
             Thanks to Tom Browder for testing and collaborating!   

2021-01-10 - Support of colorless output. Color schemes replaced with
             themes, responsible for output. 
2021-02-23 - Time zone parameter in rst_almanac.pl script, user locale auto-detection
             fixes (failed on some systems), documntation fixes
2021-03-08 - Bug fix in Rise/Set routines. On rare occasions hourly iteration could
             cross midnight and set time was missed. Documentation texts corrected, 
             thanks to Tom Browdy.  
2021-05-08 - Sidereal time: result reduced to proper range
2022-01-31 - Module for calculation time lunar phases rewritten in order to fix a 
             serious bug. Now we use a simpler but more reliable algorithm from 
             "Astronomy with Your PC" by Peter Duffett-Smith.


script/phases.pl
#!/perl

use 5.22.0;
use strict;
no warnings qw/experimental/;
use feature qw/switch/;

use utf8;
use FindBin qw/$Bin/;
use lib ("$Bin/../lib");
use Getopt::Long qw/GetOptions/;

use Pod::Usage qw/pod2usage/;
use DateTime;
use Term::ANSIColor;
use Readonly;
use Astro::Montenbruck::Utils::Helpers
    qw/parse_datetime current_timezone local_now/;
use Astro::Montenbruck::Time qw/cal2jd jd2unix/;
use Astro::Montenbruck::Lunation qw/:all/;
use Astro::Montenbruck::Utils::Theme;

my $help  = 0;
my $man   = 0;
my $date  = local_now()->truncate( to => 'hour' )->strftime('%F');
my $tzone = current_timezone();
my @place;
my $theme;

# Parse options and print usage if there is a syntax error,
# or if usage was explicitly requested.
GetOptions(
    'help|?'  => \$help,
    'man'     => \$man,
    'date:s'  => \$date,
    'theme:s' =>
        sub { $theme = Astro::Montenbruck::Utils::Theme->create( $_[1] ) },
    'no-colors' =>
        sub { $theme = Astro::Montenbruck::Utils::Theme->create('colorless') }
    ,
    'timezone:s' => \$tzone,
) or pod2usage(2);

pod2usage(1)               if $help;
pod2usage( -verbose => 2 ) if $man;

# Initialize default options

$theme //= Astro::Montenbruck::Utils::Theme->create('dark');
my $scheme = $theme->scheme;

my $dt
    = parse_datetime($date)->truncate( to => 'hour' )->set_time_zone($tzone);

$theme->print_data( 'Date',      $dt->strftime('%F'), title_width => 14 );
$theme->print_data( 'Time Zone', $tzone,              title_width => 14 );
say '';

my $ut          = $dt->clone->set_time_zone('UTC');
my @lunar_month = lunar_month( cal2jd( $ut->year, $ut->month, $ut->day ) );
for my $q (@lunar_month) {
    my $dt
        = DateTime->from_epoch( epoch => jd2unix( $q->{jd} ) )
        ->set_time_zone($tzone);
    my $mark = $q->{current} ? '*' : ' ';
    my $data = sprintf( '%s %s', $dt->strftime('%F %T'), $mark );
    $theme->print_data(
        $q->{type}, $data,
        title_width => 14,
        highlited   => $q->{current}
    );
}
say '';

__END__

=pod

=encoding UTF-8

=head1 NAME

phases — calculate date/time of principal lunar phases around a date.

=head1 SYNOPSIS

  phases [options]

=head1 OPTIONS

=over 4

=item B<--help>

Prints a brief help message and exits.

=item B<--date>

Date, either a I<calendar entry> in format C<YYYY-MM-DD>, or a floating-point I<Julian Day>:

  --date=2019-06-08 # calendar date
  --date=2438792.99 # Julian date

=item B<--timezone>

Time zone name, e.g.: C<EST>, C<UTC>, C<Europe/Berlin> etc. 
or I<offset from Greenwich> in format B<+HHMM> / B<-HHMM>, like C<+0300>.

    --timezone=CET # Central European Time
    --timezone=EST # Eastern Standard Time
    --timezone=UTC # Universal Coordinated Time
    --timezone=GMT # Greenwich Mean Time, same as the UTC
    --timezone=+0300 # UTC + 3h (eastward from Greenwich)
    --timezone="Europe/Moscow"

By default, a local timezone.

Please, note: Windows platform may not recognize some time zone names, like C<MSK>.
In such cases use I<offset from Greenwich> format, as described above.


=item B<--theme>: color theme

=over

=item * 

B<dark> (default): for dark consoles

=item * 

B<light>: for light consoles

=item * 

B<colorless>: without colors, for terminals that do not support ANSI color codes

=back

=item B<--no-colors>: do not use colors, same as C<--theme=colorless>

=back

=head1 DESCRIPTION

B<phases> Computes lunar phases around a date. Current phase (from perspective of the query date) 
is highlited and marked with asterisk.

For instance,
 
  Date          :  2021-02-22
  
  First Quarter :  2021-02-19 21:48:11 *
  Full Moon     :  2021-02-27 11:20:33

means that the query date, Feb 22, belongs to the First Quarter, while the next phase will start on Feb 27. 


=cut
